### ISSUE TITLE
expose fixed_point to_rep at cnl scope

### ISSUE THREAD

**Issue by vanossj** [johnmcfarlane/cnl#401]
```c++
auto x = cnl::fixed_point<int, -12>(1.23);
auto a = cnl::to_rep(x); //doesn't work
auto b = cnl::_impl::to_rep(x); //works
```
https://godbolt.org/z/hIB1YM

`from_rep<fixed_point, T>` is in cnl namespace, could `to_rep` be as well?

**Comment by johnmcfarlane** [comment]
Both `cnl::from_rep` and `cnl::to_rep` are function objects whereas `cnl::_impl::from_rep` and `cnl::_impl::to_rep` are functions. 

The functions are helpers for calling for the function objects and are not part of the public API. You can achieve the same thing using the function objects but it's a little more verbose, e.g. [`cnl::to_rep<cnl::fixed_point<int, -12>>{}(1.23)`](https://godbolt.org/z/fvA62Q).

A major reason for not making the functions part of the public API is that I can't think of a good naming scheme that separates the function objects from the functions. Any suggestions?

**Comment by vanossj** [comment]
Thanks for showing me that there are both functors and functions. I can see the utility of both but in my case I'm going to go with functions, the syntax of instantly invoked functors was somewhat distracting.

Naming is definitely the hardest part of software.
Here are two possible solutions, neither of them great.

Good luck, Thanks for the library.

### Option 1
* Following the standard library convetion of `std::make_unique`, `std::make_array`, ..., put functions in the `cnl` namespace.
* functors could be in another namespace `cnl::functors`

Cons:
* This subtle breaks the current api for the only benefit of not using the _impl namespace

### Option 2
from_rep/to_rep are inverse operations, that is not a trait of std::make_unique and ilk so no need to follow that convention.
* Put from_rep/to_rep functions as static in cnl::fixed_point and similar class scope
* Keep functors  in cnl scope

Cons:
* functions are behind different class namespaces instead of a single namespace

**Comment by johnmcfarlane** [comment]
Re. option 1, `std::hash` is in the `std` namespace but it's a function object so I'm not sure there's a convention there necessarily. I'd really like a function like `std::hash(any_old_object)` which did the same thing as `_impl::to_rep` and instantiated the appropriate functor but for now I'm waiting for the really compelling reason to expose such a function in the public API. 

Re. option 2, this does not seem like something that should be a member of the class. For starters, fundamental integers and 3rd-party libraries cannot just have member functions added to them. If every function associated with a type as made a member of that type, it would be a horrible mess. Think `std::string` or Java. (Rule #44 from [this guide](http://www.gotw.ca/publications/c++cs.htm) explains some more what is wrong with this approach.

More generally, I'm not convinced that `to_rep` is needed by users. What may actually be desired is a function that retrieves the innermost integer in the type -- a recursive `to_rep` if you like. Either way, I suspect that the name and the use may arrive at the same time.

Can I ask what your intended use case is?

**Comment by vanossj** [comment]
Basically serializing/deserializing data.

I get an integral number from a config file and convert it to fixed_point with from_rep.

**Comment by johnmcfarlane** [comment]
I see. That does sound like something for which the recursive `to_rep` would be appropriate.

Is it binary? `fixed_point<>` is [trivially copyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable) so you can just fread/fwrite it directly.

**Comment by hbe72** [comment]
Reading some of your other discussions I understand what you are trying to do in the context of VHDL design. I do the same thing. So here is my use-case ;)

1) I have the design for an algorithm in on float (or double) precision. Typically having that algorithm as a part of bigger Monte Carlo simulation. Maybe FFT as part of an OFDM radio system or a IIR filter as part of audio equalizer as an examples.
2) With CNL I convert that algorithm to fixed_point using necessary precision so that the algorithm still functions "well enough". The required precision of all float variables are defined by running sufficient amount of simulations (or measuring error by other means) in "all" imaginable signal scenarios. 
3) I capture the input and output of the algorithm and serialise that input to file. I may do this from multiple independent Monte Carlo runs. to_rep is needed for serialising those to file(s) as integers (or hex).
4) Give the algorithm design and the "golden references" to VHDL expert (or DSP expert) who is going to implement that algorithm block for HW. That previously captured data serves as golden reference to verify bit-exactness of the VHDL implementation. In VHDL code he manually does all the manual bit shifting (>> <<) to get fixed_point calculations correct. All that stuff what CNL automizes beautifully based on the datatype. 
5) Sometimes algorithms are so long that just input and output golden references are not enough. For initial debugging of the VHDL implementation I may serialise all or subset of the algorithms fixed_point variables for reference. Maybe from different stages of the algorithm if that is possible.
6) Or even better, the VHDL expert gets my simulation code, and the algorithm implemented with CNL, which he can debug in lockstep with his VHDL code and compare visually the results. Minor inconvenience is that he need to see the integer (or hex) representation, which can be buried quite deep in the composite datatype.

CNL is the perfect tool for algorithm designer and for documentation of the designers intent and getting those golden references. Unfortunately VHDL and DSP processor tools do not understand (yet) the datatypes composited with CNL, thus we have to communicate to those poor designers with integers, thus to_rep() and from_rep(). Personally I do not mind additional syntax, function objects or functions, as long as I get the thing done.

There still are businesses left where this is done ;) All where the energy efficiency is of utmost importance. All the modem designers in the world do this (basestations, radio chipsets, ...), video codec design, audio signal processing, basically all the silicon IP houses do this. In everything which is "hardware accelerated" to save power or achieve the processing speed. 

I used to have my own fixed point library to do all this until I found fixed_point which later became CNL.
